src/ObjCRuntime/Runtime.cs

```diff
@@ -1735,11 +1737,23 @@ internal static IntPtr GetProtocolForType (Type type)
 			throw new ArgumentException ($"'{type.FullName}' is an unknown protocol");
 		}
 
-		[BindingImpl (BindingImplOptions.Optimizable)]
 		internal static bool IsUserType (IntPtr self)
 		{
 			var cls = Class.object_getClass (self);
+			if (!usertype_cache.TryGetValue (cls, out var result)) {
+				result = SlowIsUserType (cls);
```

Comment:
Just a note, if this is required to be very fast, you can avoid double hashing:

[GetValueOrAddDefault](https://source.dot.net/#System.Private.CoreLib/CollectionsMarshal.cs,bb77eec88508f027,references) can be used to get a ref to the slot in the hash bucket, and it can be assigned direct.